### PR TITLE
Add Support for pgvector Extension

### PR DESCRIPTION
**Pull request by KellyRousselHoomano**
This pull request aims to enhance sqlacodegen's capabilities, allowing users to seamlessly integrate their PostgreSQL databases with pgvector for advanced data types such as 'Vector.' The feature caters to the increasing demand for efficient handling of embeddings in the context of Large Language Models (LLMs) and retrieval tools.

### CODE REVIEW

**Comment by KellyRousselHoomano** [comment]
**Changes Made:**
Created a dedicated branch: [feature-pgvector](https://github.com/hoomano/sqlacodegen/tree/feature-pgvector).
Followed a similar process employed for previous extensions such as "citext" or "geoalchemy2" to enable support for the pgvector extension.
Verified successful installation using:
```bash
pip install git+https://github.com/hoomano/sqlacodegen.git@feature-pgvector#egg=sqlacodegen\[pgvector\]
```

**Issue:**
The current issue is evident when attempting to export PostgreSQL database models using sqlacodegen, resulting in the following warning:

```python
/Users/kellyroussel/anaconda3/lib/python3.10/site-packages/sqlacodegen/cli.py:81: SAWarning: Did not recognize type 'vector' of column 'embedding'
  metadata.reflect(engine, schema, not args.noviews, tables)
```

File: CHANGES.rst [hunk 1]
```diff
@@ -1,6 +1,9 @@
 Version history
 ===============
 
+**3.0.0rc4**
```
Comment (agronholm) [inline CHANGES.rst:4]: The release version is only added on release.
```suggestion
**UNRELEASED**
```

File: CHANGES.rst [hunk 2]
```diff
@@ -1,6 +1,9 @@
...
+- Added support for pgvector extension
```
Comment (agronholm) [inline CHANGES.rst:5]:
```suggestion
- Added support for the `pgvector` extension
```

**Comment by agronholm** [comment]
You'll need to add the other suggested change too (I can't do it for you as you prevented me from modifying the branch).

**Comment by agronholm** [comment]
Alright, if you're okay with me merging this now, then remove the draft status please.

**Comment by KellyRousselHoomano** [comment]
@agronholm Thank you for that. Problem is I'm not sure this is working ! 

I have verified that pgvector is correctly installed when using the command:
```
pip install git+https://github.com/hoomano/sqlacodegen.git@feature-pgvector#egg=sqlacodegen\[pgvector\]
```
However, despite successful installation, running the sqlacodegen command line to export PostgreSQL database models results in the following warning:
```
sqlacodegen/cli.py:81: SAWarning: Did not recognize type 'vector' of column 'embedding'
  metadata.reflect(engine, schema, not args.noviews, tables)
```

I don't know if I'm wrong with the installation command (`egg=sqlacodegen\[pgvector\]` ?) or if this is not enough to enable pgvector in the right way?

**Comment by agronholm** [comment]
I thought you already tried it before sending this PR. Did you check if `pgvector` actually gets installed?

**Comment by agronholm** [comment]
Adding `#egg=...` to an install URL is not kosher. If you want to do a git-based install, the syntax for that is `pip install sqlacodegen[pgvector]@git+https://github.com/hoomano/sqlacodegen.git@feature-pgvector`. I can confirm that this does install the `pgvector` extra.

**Comment by KellyRousselHoomano** [comment]
Thank you for the correct syntax. I can also confirm the pgvector extension is installed correctly. 
But then running `sqlacodegen ...` it doesn't find type `Vector`. I may either be pointing to the wrong `sqlacodegen` or the PR is not working and I have no idea why 🤔

**Comment by agronholm** [comment]
Have you managed to get it work on a local checkout? If importing `pgvector` enough?

**Comment by agronholm** [comment]
It looks like `pgvector` is a namespace package and doesn't actually contain anything that would register classes with any PostgreSQL driver.

**Comment by agronholm** [comment]
What if you change it to `import pgvector.sqlalchemy`?

**Comment by agronholm** [comment]
You could also try to add:
```python
import pgvector.asyncpg
import pgvector.psycopg
import pgvector.psycopg2
```

**Comment by KellyRousselHoomano** [comment]
I managed to add:
```
import pgvector.sqlalchemy
import pgvector.psycopg2
```
Those are correctly imported but I still face the same error when running sqlacodegen command...

**Comment by agronholm** [comment]
And you're using the `psycopg2` driver to connect to your postgresql db?

**Comment by agronholm** [comment]
According to the README, these types need to be explicitly registered with a connection :P

**Comment by agronholm** [comment]
Too bad `pgvector` doesn't auto-register its types in the SQLAlchemy dialects like GeoAlchemy2 does.

**Comment by KellyRousselHoomano** [comment]
You seem to understand what's the problem but I don't get it ! What part of the README are you referring? Is there anything I can do to implement what's missing?

**Comment by agronholm** [comment]
Scratch that part about registering types on a connection. What needs to happen is for `pgvector-python` to insert its own types to SQLAlchemy's type registries on import, like [GeoAlchemy2 does](https://github.com/geoalchemy/geoalchemy2/blob/master/geoalchemy2/types/__init__.py#L382). You should probably make a PR against `pgvector-python` to make that happen.

**Comment by KellyRousselHoomano** [comment]
PR submitted to [pgvector-python repo](https://github.com/pgvector/pgvector-python/pull/45) 🤞

**Comment by agronholm** [comment]
Awesome, now we just need to wait for a release.

File: src/sqlacodegen/cli.py [hunk 3]
```diff
@@ -18,6 +18,12 @@
 except ImportError:
     geoalchemy2 = None
 
+try:
+    import pgvector.sqlalchemy
+except ImportError:
+    pgvector = None
+    print("Import error")
```
Comment (agronholm) [inline src/sqlacodegen/cli.py:25]:
```suggestion
    pass
```

**Comment by agronholm** [comment]
I made the necessary changes on my own and credited you in the changelog.

**Comment by KellyRousselHoomano** [comment]
Ow thank you !!

**Comment by pezafar** [comment]
Hi !

Thank you @agronholm and @KellyRousselHoomano for the project and the feature, this is much appreciated :)

I have a problem when installing from github, I still have the issue
`
xxx/xxx/lib/python3.10/site-packages/sqlacodegen/cli.py:85: SAWarning: Did not recognize type 'vector' of column 'xxx'
`
when using the cli. 

To install I ran `pip install "sqlacodegen[pgvector]@git+https://github.com/agronholm/sqlacodegen.git" ` but I'm not sure wether the optional dependency is installed or not. `pgvector` is well imported as I have `Using pgvector 0.2.4` but I can't figure why I still have this issue. 

(Although when installing from the @KellyRousselHoomano fork everything seemed to work)

Thanks !

**Comment by agronholm** [comment]
@KellyRousselHoomano is it working for you?

**Comment by KellyRousselHoomano** [comment]
Hi,
Indeed it's working for me on Hoomano's feature-pgvector branch (`pip install pip install "sqlacodegen[pgvector]@git+https://github.com/hoomano/sqlacodegen.git@feature-pgvector"`) but not with `pip install "sqlacodegen[pgvector]@git+https://github.com/agronholm/sqlacodegen.git"`

Trying to understand what's the difference...

**Comment by KellyRousselHoomano** [comment]
@agronholm on this repo master branch `src/sqlacodegen/cli.py` (lines 21 - 24) import is: 
```python
try:
    import pgvector
except ImportError:
    pgvector = None
```

While in my PR (and on Hoomano's fork branch) it is:

```python
try:
    import pgvector.sqlalchemy
except ImportError:
    pgvector = None
```

I think this might be the issue !

**Comment by agronholm** [comment]
Indeed. I've pushed a fix now.

**Comment by agronholm** [comment]
There's a new release out now.

**Comment by pezafar** [comment]
All good with `3.0.0rc5`, thanks !

**Pull request closed without merging**